File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/bike/ReserveBikeUseCase.java
package com.lattis.ellipse.domain.interactor.bike;


import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.model.Bike;
import com.lattis.ellipse.domain.model.Ride;
import com.lattis.ellipse.domain.repository.BikeRepository;

import javax.inject.Inject;

import io.reactivex.Observable;

public class ReserveBikeUseCase extends UseCase<Ride>{

    private BikeRepository bikeRepository;
    private Bike bike;
    private boolean by_scan;
    private double latitude;
    private double longitude;

    @Inject
    public ReserveBikeUseCase(ThreadExecutor threadExecutor,
                              PostExecutionThread postExecutionThread,
                              BikeRepository bikeRepository) {
        super(threadExecutor, postExecutionThread);
        this.bikeRepository = bikeRepository;
    }


    public ReserveBikeUseCase withBike(Bike bike) {
        this.bike = bike;
        return this;
    }

    public ReserveBikeUseCase withScanStatus(boolean by_scan) {
        this.by_scan = by_scan;
        return this;
    }

    public ReserveBikeUseCase withLatitude(double latitude) {
        this.latitude = latitude;
        return this;
    }

    public ReserveBikeUseCase withLongitude(double longitude) {
        this.longitude = longitude;
        return this;
    }

    @Override
    protected Observable<Ride> buildUseCaseObservable() {
        return bikeRepository.bookBike(bike, by_scan, latitude, longitude);
    }
}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/bike/ReserveBikeUseCase.kt
package com.lattis.domain.usecase.bike

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.BikeRepository
import com.lattis.domain.usecase.base.UseCase
import com.lattis.domain.models.Bike
import com.lattis.domain.models.Ride
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class ReserveBikeUseCase @Inject
constructor(threadExecutor: ThreadExecutor,
            postExecutionThread: PostExecutionThread,
            private val bikeRepository: BikeRepository
) : UseCase<Ride>(threadExecutor, postExecutionThread) {

    private lateinit var bike: Bike
    private var by_scan = false
    private var latitude:Double? = null
    private var longitude:Double? = null
    private var device_token:String?=null
    private var pricing_option_id:Int?=null

    fun withBike(bike: Bike): ReserveBikeUseCase {
        this.bike = bike
        return this
    }

    fun withScanStatus(by_scan: Boolean): ReserveBikeUseCase {
        this.by_scan = by_scan
        return this
    }

    fun withLatitude(latitude: Double?): ReserveBikeUseCase {
        this.latitude = latitude
        return this
    }

    fun withLongitude(longitude: Double?): ReserveBikeUseCase {
        this.longitude = longitude
        return this
    }

    fun withDeviceToken(device_token:String): ReserveBikeUseCase {
        this.device_token = device_token
        return this
    }

    fun withPricingOptionId(pricing_option_id:Int?): ReserveBikeUseCase {
        this.pricing_option_id = pricing_option_id
        return this
    }

    override fun buildUseCaseObservable(): Observable<Ride> {
        return bikeRepository.bookBike(bike, by_scan, latitude, longitude,device_token!!,pricing_option_id)
    }
}